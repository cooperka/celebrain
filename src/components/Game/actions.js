// --- Action types

export const actionTypes = {
  TOGGLE_GAME_STATE: 'TOGGLE_GAME_STATE',
  RESTART: 'RESTART',

  SET_ONLY_FIRST_NAMES: 'SET_ONLY_FIRST_NAMES',
};

// --- Action creators

export function setGameState(state) {
  return { type: actionTypes.TOGGLE_GAME_STATE, payload: { state } };
}

export function restart() {
  return { type: actionTypes.RESTART };
}

// --- Settings

export function setOnlyFirstNames(isChecked) {
  return { type: actionTypes.SET_ONLY_FIRST_NAMES, payload: { isChecked } };
}
