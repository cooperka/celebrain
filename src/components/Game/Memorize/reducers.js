import { actionTypes } from './actions';

const initialState = {
  currIndex: 0,
  numItems: 1,
};

export function memorizeReducer(state = initialState, action = {}) {
  switch (action.type) {
    case actionTypes.MEMORIZE_NEXT:
      return {
        ...state,
        currIndex: state.currIndex + 1,
      };

    default:
      return state;
  }
}
