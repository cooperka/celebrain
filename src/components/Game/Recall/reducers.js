import Immutable from 'immutable';

import { actionTypes as gameActionTypes } from '../actions';
import { actionTypes } from './actions';

const initialState = {
  currIndex: 0,

  inputText: '',
  inputs: Immutable.List(),
};

export function recallReducer(state = initialState, { type, payload } = {}) {
  switch (type) {
    case actionTypes.RECALL_NEXT:
      return {
        ...state,
        currIndex: state.currIndex + 1,
        inputText: initialState.inputText,
        inputs: state.inputs.push(state.inputText),
      };

    case actionTypes.CHANGED_INPUT_TEXT:
      return {
        ...state,
        inputText: payload.text,
      };

    case gameActionTypes.RESTART:
      return initialState;

    default:
      return state;
  }
}
