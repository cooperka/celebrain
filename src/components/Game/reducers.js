import Immutable from 'immutable';

import { getRandomOrder } from '../../utils/celeb-utils';
import { actionTypes } from './actions';

export const gameState = {
  INTRO: 'INTRO',
  MEMORIZE: 'MEMORIZE',
  RECALL: 'RECALL',
  RESULTS: 'RESULTS',
};

const initialGameState = {
  currState: gameState.INTRO,
  imageOrder: undefined,

  // --- Settings.
  numCelebs: 5,
  onlyFirstNames: false,
  popularityGroups: Immutable.List([true, false, false, false]),
};

export function gameReducer(state = initialGameState, { type, payload } = {}) {
  switch (type) {
    case actionTypes.TOGGLE_GAME_STATE: {
      const isNewGame = payload.state === gameState.MEMORIZE;

      return {
        ...state,
        currState: payload.state,
        imageOrder: isNewGame ? getRandomOrder(state.numCelebs, state.popularityGroups) : state.imageOrder,
      };
    }

    case actionTypes.SET_NUM_CELEBS:
      return {
        ...state,
        numCelebs: Math.max(payload.numCelebs, 1),
      };

    case actionTypes.SET_ONLY_FIRST_NAMES:
      return {
        ...state,
        onlyFirstNames: payload.isChecked,
      };

    case actionTypes.SET_POPULARITY:
      return {
        ...state,
        popularityGroups: state.popularityGroups.set(payload.groupIndex, payload.isChecked),
      };

    case actionTypes.RESTART:
      return initialGameState;

    default:
      return state;
  }
}
