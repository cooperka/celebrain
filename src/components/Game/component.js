// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';

import Intro from './Intro';
import Memorize from './Memorize';
import Recall from './Recall';
import Results from './Results';
import { gameState } from './reducers';

import imageData from '../../celebs.json';
import './styles.css';

type Props = {
  currState: any,
};

const gameComponent = {
  [gameState.INTRO]: Intro,
  [gameState.MEMORIZE]: Memorize,
  [gameState.RECALL]: Recall,
  [gameState.RESULTS]: Results,
};

class Game extends Component<Props> {

  static mapStateToProps(state) {
    return {
      currState: state.game.currState,
    };
  }

  render() {
    const { currState } = this.props;
    const GameComponent = gameComponent[currState];

    return (
      <div className="Game page">
        <GameComponent imageData={imageData} />
      </div>
    );
  }

}

// $FlowFixMe
export default connect(Game.mapStateToProps)(Game);
