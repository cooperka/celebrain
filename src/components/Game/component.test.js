// @flow

import R from 'ramda';
import React from 'react';
import renderer from 'react-test-renderer';

import Game from './component';

import { gameState } from './reducers';

jest.mock('./Intro/component', () => 'Intro');
jest.mock('./Memorize/component', () => 'Memorize');
jest.mock('./Recall/component', () => 'Recall');
jest.mock('./Results/component', () => 'Results');

describe('renders in state', () => {
  R.flip(R.forEachObjIndexed)(gameState, (state) => {
    it(state, () => {
      const tree = renderer.create(<Game currState={state} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
