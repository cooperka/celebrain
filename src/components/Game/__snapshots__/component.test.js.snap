// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders in state INTRO 1`] = `
<div
  className="Game page"
>
  <Intro
    imageData={
      Array [
        Object {
          "filename": "Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "id": 1806,
          "image": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/1c/Arnold_Schwarzenegger_in_Film_Festival.jpg/800px-Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "pageviews": 516051,
          "title": "Arnold Schwarzenegger",
        },
      ]
    }
  />
</div>
`;

exports[`renders in state MEMORIZE 1`] = `
<div
  className="Game page"
>
  <Memorize
    imageData={
      Array [
        Object {
          "filename": "Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "id": 1806,
          "image": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/1c/Arnold_Schwarzenegger_in_Film_Festival.jpg/800px-Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "pageviews": 516051,
          "title": "Arnold Schwarzenegger",
        },
      ]
    }
  />
</div>
`;

exports[`renders in state RECALL 1`] = `
<div
  className="Game page"
>
  <Recall
    imageData={
      Array [
        Object {
          "filename": "Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "id": 1806,
          "image": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/1c/Arnold_Schwarzenegger_in_Film_Festival.jpg/800px-Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "pageviews": 516051,
          "title": "Arnold Schwarzenegger",
        },
      ]
    }
  />
</div>
`;

exports[`renders in state RESULTS 1`] = `
<div
  className="Game page"
>
  <Results
    imageData={
      Array [
        Object {
          "filename": "Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "id": 1806,
          "image": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/1c/Arnold_Schwarzenegger_in_Film_Festival.jpg/800px-Arnold_Schwarzenegger_in_Film_Festival.jpg",
          "pageviews": 516051,
          "title": "Arnold Schwarzenegger",
        },
      ]
    }
  />
</div>
`;
